File: xrpa-orchestrator/lib/native/cs/dataset/Runtime/reconciler/ObjectCollectionIndexedBinding.cs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System.Collections.Generic;

namespace Xrpa {

  public interface IIndexBoundType<ObjectAccessorType, ReconciledType>
    where ObjectAccessorType : ObjectAccessorInterface, new()
    where ReconciledType: class, IDataStoreObjectAccessor<ObjectAccessorType> {

    bool AddXrpaBinding(ReconciledType obj);
    void RemoveXrpaBinding(ReconciledType obj);

    void TickXrpa() {}

    void WriteDSChanges();
    void ProcessDSMessage(int messageType, int timestamp, MemoryAccessor msgAccessor);
    void ProcessDSUpdate(ulong fieldsChanged);
    void ProcessDSDelete() {}
  }

  public class ObjectCollectionIndexedBinding<ObjectAccessorType, ReconciledType, IndexFieldType, LocalType> : ObjectCollectionIndex<ObjectAccessorType, ReconciledType, IndexFieldType>
    where ObjectAccessorType : ObjectAccessorInterface, new()
    where ReconciledType : class, IDataStoreObjectAccessor<ObjectAccessorType>
    where IndexFieldType : System.IEquatable<IndexFieldType>
    where LocalType : class, IIndexBoundType<ObjectAccessorType, ReconciledType> {

    public ObjectCollectionIndexedBinding(ulong inboundFieldMask) : base() {
      _inboundFieldMask = inboundFieldMask;
    }

    public void AddLocalObject(IndexFieldType indexValue, LocalType localObj) {
      // add local object to lookup
      if (!_localObjects.TryGetValue(indexValue, out var localObjects)) {
        localObjects = new List<LocalType>();
        _localObjects[indexValue] = localObjects;
      }
      localObjects.Add(localObj);

      // check if there is already a reconciled object for this index value and bind to it if it
      // exists
      var reconciledObjects = GetIndexedObjects(indexValue);
      foreach (var reconciledObj in reconciledObjects) {
        if (localObj.AddXrpaBinding(reconciledObj)) {
          var id = reconciledObj.GetXrpaId();
          _boundLocalObjects[id].Add(localObj);
          localObj.ProcessDSUpdate(_inboundFieldMask);
        }
      }
    }

    public void RemoveLocalObject(IndexFieldType indexValue, LocalType localObj) {
      // remove local object from lookup
      if (_localObjects.TryGetValue(indexValue, out var localObjects)) {
        localObjects.Remove(localObj);
        if (localObjects.Count == 0) {
          _localObjects.Remove(indexValue);
        }
      }

      // unbind local object from reconciled object
      var reconciledObjects = GetIndexedObjects(indexValue);
      foreach (var reconciledObj in reconciledObjects) {
        localObj.RemoveXrpaBinding(reconciledObj);

        var id = reconciledObj.GetXrpaId();
        if (_boundLocalObjects.TryGetValue(id, out var localBoundObjects)) {
          localBoundObjects.Remove(localObj);
          if (localBoundObjects.Count == 0) {
            _boundLocalObjects.Remove(id);
          }
        }
      }
    }

    public override void OnCreate(ReconciledType reconciledObj, IndexFieldType indexValue) {
      base.OnCreate(reconciledObj, indexValue);

      // bind local objects to reconciled object
      var id = reconciledObj.GetXrpaId();
      if (_localObjects.TryGetValue(indexValue, out var localObjects)) {
        foreach (var localObj in localObjects) {
          if (localObj.AddXrpaBinding(reconciledObj)) {
            if (!_boundLocalObjects.TryGetValue(id, out var localBoundObjects)) {
              localBoundObjects = new List<LocalType>();
              _boundLocalObjects[id] = localBoundObjects;
            }
            localBoundObjects.Add(localObj);
            localObj.ProcessDSUpdate(_inboundFieldMask);
          }
        }
      }
    }

    public override void OnDelete(ReconciledType reconciledObj, IndexFieldType indexValue) {
      var id = reconciledObj.GetXrpaId();

      if (_boundLocalObjects.TryGetValue(id, out var localBoundObjects)) {
        foreach (var localObj in localBoundObjects) {
          localObj.ProcessDSDelete();
        }
      }

      base.OnDelete(reconciledObj, indexValue);

      // unbound local objects from reconciled object
      if (localBoundObjects != null) {
        foreach (var localObj in localBoundObjects) {
          localObj.RemoveXrpaBinding(reconciledObj);
        }
        _boundLocalObjects.Remove(id);
      }
    }

    public void Tick() {
      foreach (var (id, localObjects) in _boundLocalObjects) {
        foreach (var localObj in localObjects) {
          localObj.TickXrpa();
        }
      }
    }

    public void WriteChanges(DSIdentifier id) {
      if (_boundLocalObjects.TryGetValue(id, out var localBoundObjects)) {
        foreach (var localObj in localBoundObjects) {
          localObj.WriteDSChanges();
        }
      }
    }

    public void ProcessUpdate(DSIdentifier id, ulong fieldsChanged) {
      if (_boundLocalObjects.TryGetValue(id, out var localBoundObjects)) {
        foreach (var localObj in localBoundObjects) {
          localObj.ProcessDSUpdate(fieldsChanged);
        }
      }
    }

    public void ProcessMessage(DSIdentifier id, int messageType, int timestamp, MemoryAccessor msgAccessor) {
      if (_boundLocalObjects.TryGetValue(id, out var localBoundObjects)) {
        foreach (var localObj in localBoundObjects) {
          localObj.ProcessDSMessage(messageType, timestamp, msgAccessor);
        }
      }
    }

    private ulong _inboundFieldMask;
    private Dictionary<IndexFieldType, List<LocalType>> _localObjects = new();
    private Dictionary<DSIdentifier, List<LocalType>> _boundLocalObjects = new();
  }

}

File: xrpa-orchestrator/lib/runtime/python/xrpa_runtime/reconciler/object_collection_indexed_binding.py
# Copyright (c) Meta Platforms, Inc. and affiliates.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from typing import Generic, TypeVar

from xrpa_runtime.reconciler.data_store_interfaces import IDataStoreObjectAccessor
from xrpa_runtime.reconciler.object_collection_index import ObjectCollectionIndex
from xrpa_runtime.utils.xrpa_types import ObjectAccessorInterface

ObjectAccessorType = TypeVar("ObjectAccessorType", bound=ObjectAccessorInterface)
ReconciledType = TypeVar(
    "ReconciledType", bound=IDataStoreObjectAccessor[ObjectAccessorType]
)
IndexFieldType = TypeVar("IndexFieldType")
LocalType = TypeVar("LocalType")


class ObjectCollectionIndexedBinding(
    ObjectCollectionIndex[ObjectAccessorType, ReconciledType, IndexFieldType],
    Generic[ObjectAccessorType, ReconciledType, IndexFieldType, LocalType],
):
    def __init__(self):
        ObjectCollectionIndex.__init__(self)
        self._local_objects = {}
        self._bound_local_objects = {}

    def add_local_object(self, index_value: IndexFieldType, local_obj: LocalType):
        # add local object to lookup
        local_objects = self._local_objects.get(index_value, None)
        if local_objects is None:
            local_objects = []
            self._local_objects[index_value] = local_objects
        local_objects.append(local_obj)

        # check if there is already a reconciled object for this index value and bind to it if it exists
        reconciled_objects = self.get_indexed_objects(index_value)
        for reconciled_obj in reconciled_objects:
            if local_obj.add_xrpa_binding(reconciled_obj):
                id = reconciled_obj.get_xrpa_id()
                local_bound_objects = self._bound_local_objects.get(id, None)
                if local_bound_objects is None:
                    local_bound_objects = []
                    self._bound_local_objects[id] = local_bound_objects
                local_bound_objects.append(local_obj)

    def remove_local_object(self, index_value: IndexFieldType, local_obj: LocalType):
        # remove local object from lookup
        local_objects = self._local_objects.get(index_value, None)
        if local_objects is not None:
            local_objects.remove(local_obj)
            if len(local_objects) == 0:
                del self._local_objects[index_value]

        # unbind local object from reconciled object
        reconciled_objects = self.get_indexed_objects(index_value)
        for reconciled_obj in reconciled_objects:
            local_obj.remove_xrpa_binding(reconciled_obj)

            id = reconciled_obj.get_xrpa_id()
            local_bound_objects = self._bound_local_objects.get(id, None)
            if local_bound_objects is not None:
                local_bound_objects.remove(local_obj)
                if len(local_bound_objects) == 0:
                    del self._bound_local_objects[id]

    def on_create(self, reconciled_obj: ReconciledType, index_value: IndexFieldType):
        ObjectCollectionIndex.on_create(self, reconciled_obj, index_value)

        # bind local objects to reconciled object
        id = reconciled_obj.get_xrpa_id()
        local_objects = self._local_objects.get(index_value, None)
        if local_objects is None:
            return

        for local_obj in local_objects:
            if local_obj.add_xrpa_binding(reconciled_obj):
                local_bound_objects = self._bound_local_objects.get(id, None)
                if local_bound_objects is None:
                    local_bound_objects = []
                    self._bound_local_objects[id] = local_bound_objects
                local_bound_objects.append(local_obj)

    def on_delete(self, reconciled_obj: ReconciledType, index_value: IndexFieldType):
        id = reconciled_obj.get_xrpa_id()
        ObjectCollectionIndex.on_delete(self, reconciled_obj, index_value)

        # unbind local objects from reconciled object
        local_bound_objects = self._bound_local_objects.get(id, None)
        if local_bound_objects is not None:
            for local_obj in local_bound_objects:
                local_obj.remove_xrpa_binding(reconciled_obj)
            del self._bound_local_objects[id]
